tjpcov/covariance_cluster_mass.py

```diff
@@ -0,0 +1,241 @@
+        cosmology object.  Derived attributes from the cosmology are set here.
+
+        Args:
+            cosmo (:obj:`pyccl.Cosmology`): Input cosmology
+        """
+        self.cosmo = cosmo
+        mass_def = ccl.halos.MassDef200m()
+        self.c = ccl.physical_constants.CLIGHT / 1000
+        self.mass_func = ccl.halos.MassFuncTinker08(cosmo, mass_def=mass_def)
+
+    def load_from_sacc(self, sacc_file, min_halo_mass):
+        """Cluster covariance has special parameters set in the SACC file. This
+        informs the code that the data to calculate the cluster covariance is
+        there.  We set extract those values from the sacc file here, and set
+        the attributes here.
+
+        Args:
+            sacc_file (:obj: `sacc.sacc.Sacc`): SACC file object, already
+            loaded.
+        """
+
+        z_tracer_type = "bin_z"
+        survey_tracer_type = "survey"
+        richness_tracer_type = "bin_richness"
+
+        survey_tracer = [
+            x
+            for x in sacc_file.tracers.values()
+            if x.tracer_type == survey_tracer_type
+        ]
+        if len(survey_tracer) == 0:
+            self.survey_tracer_nm = ""
+            self.survey_area = 4 * np.pi
+            print(
+                "Survey tracer not provided in sacc file.\n"
+                + "We will use the default value.",
+                flush=True,
+            )
+        else:
+            self.survey_area = survey_tracer[0].sky_area * (np.pi / 180) ** 2
+
+        # Setup redshift bins
+        z_bins = [
+            v
+            for v in sacc_file.tracers.values()
+            if v.tracer_type == z_tracer_type
+        ]
+        self.num_z_bins = len(z_bins)
+        self.z_min = z_bins[0].lower
+        self.z_max = z_bins[-1].upper
+        self.z_bins = np.round(
+            np.linspace(self.z_min, self.z_max, self.num_z_bins + 1), 2
+        )
+        self.z_bin_spacing = (self.z_max - self.z_min) / self.num_z_bins
+        self.z_lower_limit = max(0.02, self.z_bins[0] - 4 * self.z_bin_spacing)
+        # Set upper limit to be 40% higher than max redshift
+        self.z_upper_limit = self.z_bins[-1] + 0.4 * self.z_bins[-1]
+
+        # Setup richness bins
+        richness_bins = [
+            v
+            for v in sacc_file.tracers.values()
+            if v.tracer_type == richness_tracer_type
+        ]
+        self.num_richness_bins = len(richness_bins)
+        self.min_richness = 10 ** richness_bins[0].lower
+        self.max_richness = 10 ** richness_bins[-1].upper
+        self.richness_bins = np.round(
+            np.logspace(
+                np.log10(self.min_richness),
+                np.log10(self.max_richness),
+                self.num_richness_bins + 1,
+            ),
+            2,
+        )
+
+        self.min_mass = np.log(min_halo_mass)
+        self.max_mass = np.log(1e16)
+
+    def _build_matrix_from_blocks(self, blocks, tracers_cov):
```

Comment:
Why do you need to reimplement this?

Reply:
For the covariance of cluster mass, the tracers match cluster counts, so we need to explicitly grab the indices for the standard_types.cluster_mean_log_mass (see Line 163 for example).  

In the cluster counts class, the data type for cluster counts is hardcoded.

Reply:
I think it would be better to generalize the one in `CovarianceBuilder`. What do you think of adding a new attribute called `single_block` that is `False` by default but you set to `True` in the clusters covariance. Then, in `_build_matrix_from_blocks` you do 
```
if single_block is True:
    data_type1, data_type2 = self._tracer_types
else:
    data_type1 = data_type2 = None
```
That way you can use always `ix1 = s.indices(data_type=standard_types.cluster_mean_log_mass, tracers=tracer_comb1)`